### ISSUE TITLE
NET-1980 S2259 FP: Cross-procedural null-check with null conditional argument

### ISSUE THREAD

**Issue by andrei-epure-sonarsource** [SonarSource/sonar-dotnet#2721]
### Description

Null pointers should not be dereferenced FP 
![image](https://user-images.githubusercontent.com/38876598/66739877-02abb280-ee72-11e9-91cf-932668eff9ae.png)

`IsStringLiteral` does a null-check inside and returns false if null

This would need cross-procedural analysis to make it work.

### Repro steps

```
if (methodSymbol.IsConsoleWrite() || methodSymbol.IsConsoleWriteLine())
{
  var firstArgument = invocationSyntax.ArgumentList.Arguments.FirstOrDefault();
  if (IsStringLiteral(firstArgument?.Expression, c.SemanticModel))
  {
    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, firstArgument.GetLocation()));
  }
  return;
}
```

### Expected behavior

If `firstArgument` is null, it will be checked inside the `IsStringLiteral` method.
So it should not raise.

### Actual behavior

Raises S2259 on `firstArgument.GetLocation()`

### Related information

* SonarC# Version 7.17

**Comment by sit-md** [comment]
This also is true in another case I encountered yesterday. A developer used an extension to mimic `!string.IsNullOrEmpty(instance)` so he can write `instance.IsNotNullOrEmpty()`. This is currently not recognized by the analyzer for the same reason the OP stated.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @sit-md,

Do you have your extension parameter marked with `ValidatedNotNullAttribute` as the [rule description](https://rules.sonarsource.com/csharp/type/Bug/RSPEC-2259) suggests for custom validators? If not, I suggest to create a new issue as yours would be related to extension methods whereas this issue is related to conditional access scenario.

**Comment by pavel-mikula-sonarsource** [comment]
Highlighting line from the description:

>This would need cross-procedural analysis to make it work.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1980](https://sonarsource.atlassian.net/browse/NET-1980)

[NET-1980]: https://sonarsource.atlassian.net/browse/NET-1980?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
